### ISSUE TITLE
Use the default parameters when evaluating custom dataclass field operators?

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#6573]
```python
from dataclasses import KW_ONLY, InitVar, dataclass, field
from typing import Any, Self

from typing_extensions import dataclass_transform, override


def module_field(*, init: bool = False) -> Any:
    return field(init=init, default=None, kw_only=True)  # pylint: disable=invalid-field-call

@dataclass_transform(field_specifiers=(field, module_field,))
class _DataClassModule:
    @override
    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        dataclass(cls)

class X(_DataClassModule):
    y: InitVar[str]

    def __post_init__(self, y: str) -> None:
        pass

class Y(X):
    x: int = module_field()  # Error can be resolved by using module_field(init=False)

    @override
    def __post_init__(self, y: str) -> None:
        self.x = 1

y = Y("asdf")  # error: Argument missing for parameter "x"
print(y.x)
```

**Comment by erictraut** [comment]
Pyright already supports default arguments for field specifiers when the parameter type is annotated to be `Literal[False]` or `Literal[True]`. I think it's safe to also support this in the case where a default argument value evaluates to `Literal[False]` or `Literal[True]`. This will be included in the next release.

We should try to get this into the official typing spec in the `dataclass_transform` portion. I've made a note of this and will raise it with the typing council in due time. If we don't spec this, it won't work with other type checkers.

**Comment by NeilGirdhar** [comment]
> We should try to get this into the official typing spec in the `dataclass_transform` portion. I've made a note of this and will raise it with the typing council in due time. If we don't spec this, it won't work with other type checkers.

Good thinking, and thank you very much!!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
